src/basic/pidref.c

```diff
@@ -58,15 +62,23 @@ bool pidref_equal(PidRef *a, PidRef *b) {
                 if (a->fd_id == 0 || b->fd_id == 0)
                         return true;
         } else {
+                /* If the other side is remote, then this is not the same */
                 if (pidref_is_remote(b))
                         return false;
 
-                /* Try to compare pidfds using their inode numbers. This way we can ensure that we
-                 * don't spuriously consider two PidRefs equal if the pid has been reused once. Note
-                 * that we ignore all errors here, not only EOPNOTSUPP, as fstat() might fail due to
-                 * many reasons. */
-                if (pidref_acquire_pidfd_id(a) < 0 || pidref_acquire_pidfd_id(b) < 0)
+                /* PID1 cannot exit, hence it cannot change pidfs ids, hence no point in comparing them, we
+                 * can shortcut things */
+                if (a->pid == 1)
                         return true;
+
+                /* Try to compare pidfds using their inode numbers. This way we can ensure that we don't
+                 * spuriously consider two PidRefs equal if the pid has been reused once. Note that we ignore
+                 * all errors here, not only EOPNOTSUPP, as fstat() might fail due to many reasons. Note that
+                 * we only acquire the fd id if we already have the other one, because otherwise we'd just
+                 * acquire the same id twice. */
+                if ((a->fd_id == 0) != (b->fd_id == 0))
```

Comment:
Hmm, thinking about it more, I don't find this shortcut very compelling. When I added this initially I thought about e.g. logind/machined monitoring external process, which might be terminated already but we haven't yet released reference to it, but in the meantime another client comes over with the reused pid. Hence I figure we should not just return true here.

Of course, our hash ops and such still stores pidref as pid, so this is more like paranoia. However for an explicit compare func I don't feel like just dropping the check.

Reply:
not sure i follow? the thing is that if for both pidrefs we only have the pid to go on, and we then acquire the pidfd id for both of them to compare them, then of course they will also match since at the time of the comparison we opened *both* of the pids, and of course that's going to be the same.

hence i figured, wtf, let's just skip the step in this case, it's pointless: let's only acquire the pidfd ids here if *one* of the pidrefs has it already and the other does not.

Reply:
> the thing is that if for both pidrefs we only have the pid to go on

Not really? What I'm talking about is two pidrefs with same pid but pidfd pointing to different processes.